src/Compilers/Core/Portable/PEWriter/MethodDefinitionBase.cs

```diff
@@ -0,0 +1,182 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Reflection;
+using Microsoft.CodeAnalysis.CodeGen;
+using Microsoft.CodeAnalysis.Debugging;
+using Microsoft.CodeAnalysis.Emit;
+using Roslyn.Utilities;
+
+namespace Microsoft.Cci;
+
+internal abstract class MethodDefinitionBase : IMethodDefinition, IMethodBody
+{
+    public MethodDefinitionBase(ITypeDefinition containingTypeDefinition, ushort maxStack, ImmutableArray<byte> il)
+    {
+        ContainingTypeDefinition = containingTypeDefinition;
+        MaxStack = maxStack;
+        IL = il;
+    }
+
+    // IMethodDefinition implementation
+
+    bool IDefinition.IsEncDeleted => false;
+
+    public ITypeDefinition ContainingTypeDefinition { get; }
+
+    public abstract string Name { get; }
+
+    public bool HasBody => true;
+
+    public IMethodBody GetBody(EmitContext context) => this;
+
+    public IEnumerable<IGenericMethodParameter> GenericParameters => SpecializedCollections.EmptyEnumerable<IGenericMethodParameter>();
+
+    public bool HasDeclarativeSecurity => false;
+
+    public bool IsAbstract => false;
+
+    public bool IsAccessCheckedOnOverride => false;
+
+    public bool IsConstructor => false;
+
+    public bool IsExternal => false;
+
+    public bool IsHiddenBySignature => true;
+
+    public bool IsNewSlot => false;
+
+    public bool IsPlatformInvoke => false;
+
+    public virtual bool IsRuntimeSpecial => false;
```

Comment:
>virtual [](http://example.com/codeflow?start=11&length=7)

It looks like there are only two distinct implementations of `MethodDefinitionBase`. Given that, consider using `abstract` members in this base type and implementing in both derived types.

Reply:
I would prefer to not do that. I imagine in the future there could be more derived types and they probably won't need to override IsRuntimeSpecial and IsSpecialName. The purpose of this base class is to share code.